libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Properties.hs

```diff
@@ -479,26 +533,34 @@ genDCerts = do
         case dc of
           DCertDeleg d
             | RegKey regCred <- d ->
-                if regCred `Set.member` regCreds
+                if regCred `hasKey` regCreds
                   then pure (dcs, ss, regCreds)
-                  else pure (dc : dcs, ss, Set.insert regCred regCreds)
-            | DeRegKey deregCred <- d ->
-                if deregCred `Set.member` regCreds
-                  then
-                    insertIfNotPresent dcs (Set.delete deregCred regCreds) Nothing
+                  else pure (dc : dcs, ss, Map.insert regCred (Coin 99) regCreds) -- 99 is a NonZero Value
+            | DeRegKey deregCred <- d -> -- We can't make DeRegKey certificate if the Rewards balance is not 0
+                case Map.lookup deregCred regCreds of
+                  Nothing -> pure (dcs, ss, regCreds) -- No credential, skip making certificate
+                  Just (Coin 0) ->
+                    -- Ok to make one, rewards balance is 0
+                    insertIfNotPresent dcs (Map.delete deregCred regCreds) Nothing
                       <$> lookupPlutusScript deregCred Cert
-                  else pure (dcs, ss, regCreds)
+                  -- WHY do we look at JUST plutus scripts? FIXME
+                  Just (Coin _) -> pure (dcs, ss, regCreds) -- Reward balance is not zero, skip making certificate
             | Delegate (Delegation delegCred delegKey) <- d ->
                 let (dcs', regCreds')
-                      | delegCred `Set.member` regCreds = (dcs, regCreds)
+                      | delegCred `hasKey` regCreds = (dcs, regCreds)
                       | otherwise =
-                          (DCertDeleg (RegKey delegCred) : dcs, Set.insert delegCred regCreds)
+                          (DCertDeleg (RegKey delegCred) : dcs, Map.insert delegCred (Coin 99) regCreds)
                  in insertIfNotPresent dcs' regCreds' (Just delegKey)
-                      <$> lookupPlutusScript delegCred Cert
+                      <$> lookupPlutusScript delegCred Cert -- WHY do we look at JUST plutus scripts? FIXME ??
```

Comment:
I am not sure I follow the `FIXME` comment here, could you elaborate?

Reply:
I got rid of it.